gax/retry_policy.h

```diff
@@ -0,0 +1,119 @@
+// Copyright 2019 Google Inc.  All rights reserved
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_API_CODEGEN_GAPIC_RETRY_POLICY_H_
+#define GOOGLE_API_CODEGEN_GAPIC_RETRY_POLICY_H_
+
+#include <chrono>
+#include <memory>
+
+namespace google {
+namespace gapic {
+namespace gax {
+
+/**
+ * Define the interface for controlling how clients retry RPC operations.
+ *
+ * Idempotent (and certain non-idempotent) operations can be retried
+ * transparently to the user. However, we need to give the users enough flexiblity
+ * to control when to stop retrying.
+ *
+ * The application provides an instance of this class when the client is created.
+ */
+template <typename StatusType, typename RetryablePolicy>
```

Comment:
We discussed the template approach offline in the doc - do you think it is the best way forward? Is there benefit to implementing it this way vs just using the known `gax::Status` type? (other than that this type not currently existing, meaning we would need to implement `gax::Status` first?)

Reply:
I initially tried doing it with out, but the templating kind of just fell into place due to TDD. I'm open to removing it and adding gax::Status first. Alternatively, we can keep the templating as a source of mock types for the time being and then remove it when we have gax::Status and just use that concretely.

Reply:
SG, let's keep it in this PR but introduce gax::Status soon, and then replace.